files/private-chef-ctl-commands/key-control.rb

```diff
@@ -14,345 +14,418 @@
 # limitations under the License.
 #
 
-require "sequel"
 require "openssl"
-require "time"
+require "optparse"
+require "ostruct"
+require "chef/config"
 
 cmd_args = ARGV[3..-1]
-
-##################
-# Helper Classes #
-##################
-class Database
-  attr_reader :db
-  def initialize(running_config)
-    if running_config.nil?
-      STDERR.puts "Error: you must run chef-server-ctl reconfigure before using key commands."
-      exit 1
-    end
-
-    # Pull all the DB params from config; the database name seems to be hardcoded,
-    # however -- I found it in the postgresql.rb recipe in the private-chef cookbooks;
-    # I'm not sure why we hardcode it instead of making it configurable, but I assume
-    # there are reasons?
-    db_user = running_config['private_chef']['postgresql']['sql_user']
-    db_password = running_config['private_chef']['postgresql']['sql_password']
-    db_host = running_config['private_chef']['postgresql']['listen_address']
-
-    # Set convert_infinite_timestamps to a float so that Sequel
-    # doesn't crash out on querying an infinity timestamp.
-    # See this commit for more detail:
-    # https://github.com/jeremyevans/sequel/commit/2440e98b4a7063a90d8337e57b1c3e3d3a1cbd0b
-    @db = Sequel.connect(:adapter => 'postgres', :host => db_host,
-                         :database => 'opscode_chef', :user => db_user,
-                         :password => db_password, :convert_infinite_timestamps => :float)
-
-  end
-end
-
-class PublicKeyOutputter
-  def output_key_results(results, hide_public_keys, database)
-    results.each do |row|
-      # get expiration date from keys table as
-      # results variable is result of keys_by_name view
-      expires_at = database.db[:keys].where(:id => row[:id]).where(:key_name => row[:key_name]).first[:expires_at]
-      puts "\nkey_name: #{row[:key_name]}"
-      puts "expires_at: #{expires_at}"
-      unless hide_public_keys
-        puts "public_key:"
-        puts "#{row[:public_key]}"
+Chef::Config.from_file("/etc/opscode/pivotal.rb")
+@chef_rest = Chef::REST.new(Chef::Config[:chef_server_root])
+
+class Helper
+  # since optparse is stupid, make sure that the "argument" passed into the currently
+  # parsed argflag isn't the next argflag. if it is an argflag, they passed nothing
+  # to the current argument, so throw  the proper error.
```

Comment:
I don't think this should be necessary.  Optparse supports both optional and non-optional arguments.  From their docs:

```
      # Mandatory argument.
      opts.on("-r", "--require LIBRARY",
              "Require the LIBRARY before executing your script") do |lib|
        options.library << lib
      end

      # Optional argument; multi-line description.
      opts.on("-i", "--inplace [EXTENSION]",
              "Edit ARGV files in place",
              "  (make backup if EXTENSION supplied)") do |ext|
        options.inplace = true
        options.extension = ext || ''
        options.extension.sub!(/\A\.?(?=.)/, ".")  # Ensure extension begins with dot.
      end
```